pyface/ui/qt4/data_view/data_view_item_model.py

```diff
@@ -0,0 +1,235 @@
+
+    def parent(self, index):
+        if not index.isValid():
+            return QModelIndex()
+        parent = index.internalPointer()
+        if parent == Root:
+            return QModelIndex()
+
+        grandparent, row = self.model.index_manager.get_parent_and_row(parent)
+        return self.createIndex(row, 0, grandparent)
+
+    def index(self, row, column, parent):
+        if parent.isValid():
+            parent_index = self.model.index_manager.create_index(
+                parent.internalPointer(),
+                parent.row(),
+            )
+        else:
+            parent_index = Root
+        index = self.createIndex(row, column, parent_index)
+        return index
+
+    def rowCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            if self.model.can_have_children(row_index):
+                return self.model.get_row_count(row_index)
+        except Exception:
+            logger.exception("Error in rowCount")
+
+        return 0
+
+    def columnCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            return self.model.get_column_count(row_index) + 1
+        except Exception:
+            logger.exception("Error in columnCount")
+
+    # Data methods
+
+    def flags(self, index):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+
+        flags = Qt.ItemIsEnabled | Qt.ItemIsSelectable
+        if is_qt5 and not self.model.can_have_children(row):
+            flags |= Qt.ItemNeverHasChildren
+
+        if value_type and value_type.can_edit(self.model, row, column):
+            flags |= Qt.ItemIsEditable
+
+        return flags
+
+    def data(self, index, role=Qt.DisplayRole):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+        if not value_type:
+            return None
+
+        if role == Qt.DisplayRole:
+            if value_type.has_text(self.model, row, column):
+                return value_type.get_text(self.model, row, column)
+        elif role == Qt.EditRole:
+            if value_type.can_edit(self.model, row, column):
+                return value_type.get_editable(self.model, row, column)
+
+        return None
+
+    def setData(self, index, value, role=Qt.EditRole):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+        if not value_type:
+            return False
+
+        if role == Qt.EditRole:
+            if value_type.can_edit(self.model, row, column):
+                return value_type.set_editable(self.model, row, column, value)
+        elif role == Qt.TextRole:
+            if value_type.has_text(self.model, row, column):
+                return value_type.set_text(self.model, row, column, value)
+
```

Comment:
No, this is (and _has_ to be, or we will have signals firing twice) handled via the `value_changed` mechanisms on the model.

When the `set_value` method of the `AbstractDataModel` gets called, it will make a change to the underlying data, and that underlying data will frequently to be being listened to somehow because it may be changed in other parts of the program.  That change will trigger a `value_changed` event which in turn will cause the `dataChanged` signal to fire.  If we fired `dataChanged` here then it would end up being fired twice, once in this method, and once in response to the trait change.

In cases where we can't listen to the underlying data (such as a numpy array) then code which changes the data has to explicitly fire the `value_changed` event on every change, but that isn't a big imposition because we will want to have some way of knowing that the user made a change so that other, external, objects can listen to the data model if they want.

Reply:
Ah... I did not realize that the `set_value` on `AbstractDataModel` needs to call `value_changed`. 
So the `dataChanged.emit` is called.
`set_value` has so many responsibility... it already needs to return a boolean indicating the value has changed, why isn't that boolean used here for calling `dataChanged.emit` but instead `set_value` has to fire change event on `values_changed` as well?